app/src/main/java/com/keylesspalace/tusky/components/viewthread/ViewThreadViewModel.kt

```diff
@@ -0,0 +1,426 @@
+            status.copy(pinned = event.pinned)
+        }
+    }
+
+    private fun removeAllByAccountId(accountId: String) {
+        _uiState.updateSuccess { uiState ->
+            uiState.copy(
+                statuses = uiState.statuses.filter { viewData ->
+                    viewData.status.account.id == accountId
+                }
+            )
+        }
+    }
+
+    private fun handleStatusComposedEvent(event: StatusComposedEvent) {
+        val eventStatus = event.status
+        _uiState.updateSuccess { uiState ->
+            val statuses = uiState.statuses
+            val detailedIndex = statuses.indexOfFirst { status -> status.isDetailed }
+            val repliedIndex = statuses.indexOfFirst { status -> eventStatus.inReplyToId == status.id }
+            if (detailedIndex != -1 && repliedIndex >= detailedIndex) {
+                // there is a new reply to the detailed status or below -> display it
+                val newStatuses = statuses.subList(0, repliedIndex + 1) +
+                    eventStatus.toViewData() +
+                    statuses.subList(repliedIndex + 1, statuses.size)
+                uiState.copy(statuses = newStatuses)
+            } else {
+                uiState
+            }
+        }
+    }
+
+    private fun handleStatusDeletedEvent(event: StatusDeletedEvent) {
+        _uiState.updateSuccess { uiState ->
+            uiState.copy(
+                statuses = uiState.statuses.filter { status ->
+                    status.id != event.statusId
+                }
+            )
+        }
+    }
+
+    fun toggleRevealButton() {
+        _uiState.updateSuccess { uiState ->
+            when (uiState.revealButton) {
+                RevealButtonState.HIDE -> uiState.copy(
+                    statuses = uiState.statuses.map { viewData ->
+                        viewData.copy(isExpanded = false)
+                    },
+                    revealButton = RevealButtonState.REVEAL
+                )
+                RevealButtonState.REVEAL -> uiState.copy(
+                    statuses = uiState.statuses.map { viewData ->
+                        viewData.copy(isExpanded = true)
+                    },
+                    revealButton = RevealButtonState.HIDE
+                )
+                else -> uiState
+            }
+        }
+    }
+
+    private fun List<StatusViewData.Concrete>.getRevealButtonState(): RevealButtonState {
+        val hasWarnings = any { viewData ->
+            viewData.status.spoilerText.isNotEmpty()
+        }
+
+        return if (hasWarnings) {
+            val allExpanded = none { viewData ->
+                !viewData.isExpanded
+            }
+            if (allExpanded) {
+                RevealButtonState.HIDE
+            } else {
+                RevealButtonState.REVEAL
+            }
+        } else {
+            RevealButtonState.NO_BUTTON
+        }
+    }
+
+    private fun loadFilters() {
+        viewModelScope.launch {
```

Comment:
we again load them async but I guess there's no way around this for now, right?

Reply:
Yes, needs to be redone with filters caching.